Add read-only mode to AnimationTreeEditor plugins

SaracenOne opened the pull request:
Adds a read-only mode to the AnimationTreeEditor plugin when operating on resources embedded in foreign scenes. Currently only covers state machines, blend 1D, and blend 2D editors. Blend trees do not currently support a read-only mode due to their reliance on the built-in GraphEdit control which does not yet have the required read-only state needed to prevent edits.

Complementary PR to #63245

SaracenOne committed:
Add read-only mode to AnimationTreeEditor plugins

reduz commented on editor/plugins/animation_blend_space_1d_editor.cpp:
```diff
@@ -473,7 +479,7 @@ void AnimationNodeBlendSpace1DEditor::_update_edited_point_pos() {
 void AnimationNodeBlendSpace1DEditor::_update_tool_erase() {
 	bool point_valid = selected_point >= 0 && selected_point < blend_space->get_blend_point_count();
-	tool_erase->set_disabled(!point_valid);
+	tool_erase->set_disabled(!point_valid || blend_space_is_foreign);
```
I would just call it read_only, as I mentioned in other PRs, we also want this for locked resources for VCS eventually.

reduz commented:
Looks good to me save for minor nitpick, I think @TokageItLab should review too.

TokageItLab commented:
I have confirmed that the change to BlendSpace is prevented. But I concerned about the behavior of the BlendSpace point being canceled after it is grabbed. It would be better if it could be prevented from grabbing the point.

I am more concerned that #63245 is not working very well. I think that instead of preventing overwrites, #63245 should also cancel the user's operation, just like this PR.

SaracenOne commented:
Okay, @reduz, this has now been updated to address the issues @TokageItLab pointed out; the dragging of points in the blend spaces is no longer allowed, amongst other tweaks such as disabling a few buttons which I missed. I apologize that the read-only modes for blend spaces was so sloppy, I think that part of the code was a bit rushed on my behalf. I also went and replaced some of the duplicate code required for detecting if a resource is read_only with the new 'is_resource_read_only' function introduced in #63282

Aside from that, I've actually gone and ostensibly *finished* this PR by going ahead and adding read-only mode to the blend trees too, which did requires some minor tweaks the GraphEdit and GraphNode APIs, such as including a new 'draggable' and 'selectable' property for GraphNodes, and an option to hide the Arrange Nodes button in GraphEdit. I also introduced a new behaviour to GraphNode slots which does have accompanying documentation explaining it. The rule introduced is that slot types with negative numbers will not allow the user to create connections via the GraphEdit, which taking advantage of now allows the GraphEdit node in the BlendTreeEditor to be exist in a fully read-only state.

It also fixes a bug in the EditorSpinSplider which allowed the slider to still be moved in read-only mode.

TokageItLab commented:
Looks good except that the name of the StateMachine's Nodes is editable.

TokageItLab reviewed (approved):
Seems good.

akien-mga commented:
Thanks!
